pkg/registries/registries.go

```diff
@@ -0,0 +1,139 @@
+			set := &mirrorSet{
+				source:             entry.Source,
+				mirrorSourcePolicy: entry.MirrorSourcePolicy,
+			}
+			for _, image := range entry.Mirrors {
+				set.mirrors = append(set.mirrors, mirror{host: string(image), resolveTags: true})
+			}
+			allMirrorSets = append(allMirrorSets, *set)
+		}
+	}
+
+	return &registryConfig{sourceConfigs: mergeMirrorSets(allMirrorSets)}
+}
+
+// mergeMirrorSets consolidates duplicate entries in the given slice (based on the source) since we do not want to
+// generate multiple config files for the same source image repo. Output is sorted to ensure it is deterministic.
+func mergeMirrorSets(baseMirrorSets []mirrorSet) []mirrorSet {
+	// Map to keep track of unique mirrorSets by source
+	uniqueMirrorSets := make(map[string]mirrorSet)
+
+	for _, ms := range baseMirrorSets {
+		if existingMS, ok := uniqueMirrorSets[ms.source]; ok {
+			// If the source already exists, merge its mirrors slices
+			existingMS.mirrors = mergeMirrors(existingMS.mirrors, ms.mirrors)
+			// If the existing source's mirrorSourcePolicy conflicts, NeverContactSource is preferred
+			if existingMS.mirrorSourcePolicy == config.AllowContactingSource {
+				existingMS.mirrorSourcePolicy = ms.mirrorSourcePolicy
+			}
+			// Update the map entry
+			uniqueMirrorSets[ms.source] = existingMS
+		} else {
+			// If it does not exist, add it to the map
+			uniqueMirrorSets[ms.source] = ms
+		}
+	}
+
+	// Convert the map back to a slice with no duplicates sources
+	var result []mirrorSet
+	for _, ms := range uniqueMirrorSets {
+		result = append(result, ms)
+	}
+
+	sortMirrorSets(result)
+	return result
+}
+
+// sortMirrorSets sorts the mirrorSets and each set of underlying mirrors aplhabetically. Modifies the parameter in place
+func sortMirrorSets(mirrorSets []mirrorSet) {
+	// Sort mirrors by host alphabetically within each mirrorSet
+	for i := range mirrorSets {
+		sort.Slice(mirrorSets[i].mirrors, func(j, k int) bool {
+			return mirrorSets[i].mirrors[j].host < mirrorSets[i].mirrors[k].host
+		})
+	}
+	// Sort mirrorSets by source alphabetically
+	sort.Slice(mirrorSets, func(i, j int) bool {
+		return mirrorSets[i].source < mirrorSets[j].source
+	})
+}
+
+// mergeMirrors consolidates duplicate mirrors in the given slice (based on the host) since we do not want to
+// generate multiple entries in a single config file for the same mirror repo
+func mergeMirrors(existingMirrors, newMirrors []mirror) []mirror {
+	// Map to keep track of unique mirrors by host
+	uniqueMirrors := make(map[string]mirror)
+
+	// Iterate over existing mirrors and add them to the map
+	for _, m := range existingMirrors {
+		uniqueMirrors[m.host] = m
+	}
+	// Iterate over new mirrors
+	for _, m := range newMirrors {
+		if existingM, ok := uniqueMirrors[m.host]; ok {
+			// If the mirror already exists, check the resolveTags field. Resolving by tag is preferred over by digest.
+			if !existingM.resolveTags && m.resolveTags {
```

Comment:
If both existing and new mirrors have resolveTags set to true, would we prefer one over the other based on other attributes, such as any priority or creation time?

Reply:
Don't see why we'd need to. `mirror` structs only have 2 fields, and if both are the same them then the objects are equivalent -- it wouldn't matter which copy we keep